akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala

```diff
@@ -89,54 +89,99 @@ private[http] object HttpServerBluePrint {
-
-            val entity = createEntity(entityCreator) withSizeLimit settings.parserSettings.maxContentLength
-            push(out, HttpRequest(effectiveMethod, uri, effectiveHeaders, entity, protocol))
-          case other ⇒
-            throw new IllegalStateException(s"unexpected element of type ${other.getClass}")
+      setIdleHandlers()
+
+      def setIdleHandlers() {
+        setHandler(in, this)
+        setHandler(out, this)
+        if (upstreamPullWaiting) {
+          upstreamPullWaiting = false
+          pull(in)
         }
       }
-      setHandler(in, idle)
+
 
       def createEntity(creator: EntityCreator[RequestOutput, RequestEntity]): RequestEntity =
         creator match {
           case StrictEntityCreator(entity) ⇒ entity
-          case StreamedEntityCreator(creator) ⇒
-            val entitySource = new SubSourceOutlet[RequestOutput]("EntitySource")
-            entitySource.setHandler(new OutHandler {
-              def onPull(): Unit = pull(in)
-            })
-            setHandler(in, new InHandler {
-              def onPush(): Unit = grab(in) match {
-                case MessageEnd ⇒
-                  entitySource.complete()
-                  setHandler(in, idle)
-                case x ⇒ entitySource.push(x)
-              }
-              override def onUpstreamFinish(): Unit = completeStage()
-            })
-            creator(Source.fromGraph(entitySource.source))
+          case StreamedEntityCreator(creator) ⇒ streamRequestEntity(creator)
         }
 
-      setHandler(out, new OutHandler {
-        override def onPull(): Unit = pull(in)
-      })
+      def streamRequestEntity(creator: (Source[ParserOutput.RequestOutput, NotUsed]) => RequestEntity): RequestEntity = {
+        // stream incoming chunks into the request entity until we reach the end of it
+        // and then toggle back to "idle"
+
+        val entitySource = new SubSourceOutlet[RequestOutput]("EntitySource")
+        // optimization: re-use the idle outHandler
+        entitySource.setHandler(this)
+
+        setHandler(in, new InHandler {
+          def onPush(): Unit = {
+            grab(in) match {
+              case MessageEnd ⇒
+                entitySource.complete()
+                setIdleHandlers()
+
+              case x ⇒ entitySource.push(x)
+            }
+          }
+          override def onUpstreamFinish(): Unit = {
+            entitySource.complete()
+            completeStage()
+          }
+          override def onUpstreamFailure(ex: Throwable): Unit = {
+            entitySource.fail(ex)
+            failStage(ex)
+          }
+        })
+        setHandler(out, new OutHandler {
+          override def onPull(): Unit = {
+            // remember this until we are done with the entity
+            // so we can pass it downstream at that point
+            upstreamPullWaiting = true
+          }
+          override def onDownstreamFinish(): Unit = {
+            entitySource.complete()
```

Comment:
actually I’d need to think more about this (might be okay for this specific case though): in general not wanting the next request does not imply also not wanting the rest of the current one, so this stage completion might want to be deferred until `setIdleHandlers` tim